### ISSUE TITLE
Create Piece method to only return unblocked straight-line moves

### ISSUE THREAD

**Issue by MarkZH** [MarkZH/Genetic_Chess#64]
Instead of duplicating the logic of straight-line moves in `Board::safe_for_king()`, `Board::refresh_checking_squares()`, and other places (where the knight-simplifications were made in commits 08125cf4a0d7f1eb305a9db769fd329e4733c013, ef516a429f5e9e452c6bc1b3a4e606f906de947e, and a2333467856a92cc83ecea65a5ba5a24b50cc4d8), create a structure that takes a file, rank, and a const-ref to the board and outputs iterators that skip straight-line moves that are blocked. See [here](https://stackoverflow.com/a/36606852/2156537) for examples of how to implement iterators. This is not needed for the knight since its moves cannot be blocked (hence the commits referenced above using the `Piece::get_move_list()` method.

This may replace `Piece::get_move_list()` if the Sphere of influence gene keeps acting weird with valuing currently illegal moves over legal moves for scoring how much of the board is controlled.

**Comment by MarkZH** [comment]
Idea: adapt 

    std::array<size_t, 64> all_square_indices_attacked_by(Color player) const;

To return a reference to a new `Board` member: `std::array<std::array<bool, 64>, 2> Board::attacked_squares;` This 2D array will be indexed by `[Color][square index]`. It will be repopulated during `Board::submit_move()` after every move with a new method, `Board::refresh_attacked_squares()` maybe. Then, `Board::safe_for_king()` calls this function. `Board::refresh_checking_squares()` should be merged into the new method.

Make sure to check speed afterwards.

**Comment by MarkZH** [comment]
Solution complete in attacking_move_iterator branch.
